### PR TITLE
issue #92 added three dots svg icon and logic to show/hide quoted res…

### PR DESCRIPTION
**Pull request by michael-volynets**
…ponse

Issue #92 
Time: 18 hours (after review)
Most time I've spent on testing and thinking what to do if we forwarded HTML message.
We have this `if` statement
```
if (!Mime.resemblesMsg(decryptedBytes)) {
   ...
} else {
      renderText('Formatting...');
      const decoded = await Mime.decode(decryptedBytes);
      if (typeof decoded.html !== 'undefined') {
        await renderContent(decoded.html, false);
      } else if (typeof decoded.text !== 'undefined') {
        await renderContent(Xss.escapeTextAsRenderableHtml(decoded.text), false);
      } else {
        await renderContent(Xss.escapeTextAsRenderableHtml(decryptedContent), false);
      }
    ....
}
```
I didn't know if there could be Forwarded messages before you sent me a message to test images it was HTML message, I checked the behavior of forwarding HTML messages and realized if we forward HTML message it becomes a simple text message: 
image
So we can't forward HTML messages yet and I don't need to write any code there

Am I Correct?

### CODE REVIEW

**Comment by tomholub** [comment]
Thank you, I need to clarify something: this should be a more general solution, not just for forwarded messages. Try to send yourself an email and reply to it from Enigmail. Enigmail will quote your previous message and send it with the reply. That may be HTML, or it may be text, etc.

FlowCrypt will similarly start to quote previous message during a reply (not just forward) in the future. So that is what we are trying to address.

Basically you want to loop through the text and look for a continuous group of lines (from the end) that start with `>`, and replace these with the triple-dot, similar to what Gmail does when you reply to an email.

That means HTML situations will have to be handled as well, because other software (eg Enigmail) do that, plus FlowCrypt will also use HTML one day. You can send yourself various replies from Enigmail to test this.

**Comment by michael-volynets** [comment]
Okay, got it.

It's working now with HTML (with blockquote tag) and with plain text (tries to find '>' symbol).
but I want to test it with other email clients that have encryption.
What other clients would you recommend for testing it?

**Comment by tomholub** [comment]
I went through `flowcrypt.compatibility` emails to find a few samples:

https://mail.google.com/mail/u/2/#label/tests%2Fencrypted/WhctKJTrWNFtCxXXWRKkdrcSrfJrjGcmCLtvNgTqtkGgssRfhhTgqNnVpCXJpKXwVVWwgpG

https://mail.google.com/mail/u/2/#label/tests%2Fencrypted/WhctKJTrSJSQMNZhTbjBgNlsFDfcbcnCvJvkzTJHSPZrtRkGfKvsczSHCLvvGnRdXQCgQPV

I cannot find a plain text sample reply, you could get one by switching to plain text in enigmail, then replying. Actually it would be good to get one such email sent to `flowcrypt.compatibility` for a future automated test.

Otherwise, other than Enigmail, you could try:
 - Apple Mail + gpgmail (GPG Tools)
 - K-9 + OpenKeychain
 - ProtonMail (I'm not sure if it quotes previous replies, maybe not)

Or anything listed at https://flowcrypt.com/docs/technical/openpgp-compatibility.html that's available for download.

**Comment by michael-volynets** [comment]
I've tested it via `protonmail` and it works the same way as enigmail.

Also, I've updated the time I spent on the issue

**Comment by tomholub** [comment]
Thanks! Is it ready for second review?

**Comment by michael-volynets** [comment]
Some tests are failed, wait a while please I will check it

**Comment by michael-volynets** [comment]
Tests are passed now.
@tomholub  Sorry for the delay, I forget to move one `if ` expression

File: extension/chrome/elements/pgp_block.ts [hunk 1]
```diff
@@ -436,6 +453,28 @@ Catch.try(async () => {
...
+        `${Xss.htmlSanitizeKeepBasicTags(message)}
+          </div>`);
+    } else {
+      pgpBlock.append(`<div id="forwarded_msg" class="forwarded_msg" style="display:none">` + // xss-safe-factory
+        `${Xss.htmlSanitizeKeepBasicTags(Xss.escapeTextAsRenderableHtml(message))}
+                  </div>`);
+    }
+    $('#action_open_full').click(Ui.event.handle(async target => {
+      if ($("#forwarded_msg").css('display') === 'none') {
+        $("#forwarded_msg").css('display', 'block');
```
Comment (tomholub) [inline extension/chrome/elements/pgp_block.ts]: Could shorten these 5 lines to
```js
$("#forwarded_msg").css('display', $("#forwarded_msg").css('display') === 'none' ? 'block' : 'none');
```

File: extension/chrome/elements/pgp_block.ts [hunk 2]
```diff
@@ -292,48 +292,65 @@ Catch.try(async () => {
...
           renderableAtts.push(att);
         } else {
           publicKeys.push(att.getData().toUtfStr());
         }
       }
-      if (renderableAtts.length) {
-        renderInnerAtts(decoded.atts);
+    }
+    if (isHtml) {
+      const $message = $('<div>').html(Xss.htmlSanitize(decryptedContent)); // xss-safe-factory
```
Comment (tomholub) [inline extension/chrome/elements/pgp_block.ts]: I like how you distinguish jQuery variables with `$variable`. I think it adds some clarity and it's succinct. But it's inconsistent with the rest of the codebase, we don't use it anywhere else. Good things are worth doing consistently.

I suggest you rename these to not use the $ anymore, for consistency. Then create an issue to suggest to refactor variables in a similar fashion across the codebase, to do sometime later.

File: extension/chrome/elements/pgp_block.ts [hunk 3]
```diff
@@ -436,6 +453,28 @@ Catch.try(async () => {
     }
   };
 
+  const addSecondPart = (message: string, isHtml: boolean = false) => {
+    const pgpBlock = $("#pgp_block")
+      .append('<div id="action_open_full" class="three_dots"><img src="/img/svgs/three-dots.svg" /></div>'); // xss-safe-factory
+    if (isHtml) {
+      pgpBlock.append(`<div id="forwarded_msg" class="forwarded_msg" style="display:none">` + // xss-safe-factory
```
[the remark this answers is not in the document]
Comment (tomholub) [reply extension/chrome/elements/pgp_block.ts]: Also, it may not need both `id="forwarded_msg"` and `class="forwarded_msg"` unless there is a particular reason for it.

If you want, you could also move `style="display:none"` to css. That will shorten the code considerably, possibly fitting on one line.

File: extension/chrome/elements/pgp_block.ts [hunk 4]
```diff
@@ -436,6 +453,28 @@ Catch.try(async () => {
...
+        `${Xss.htmlSanitizeKeepBasicTags(message)}
+          </div>`);
+    } else {
+      pgpBlock.append(`<div id="forwarded_msg" class="forwarded_msg" style="display:none">` + // xss-safe-factory
```
[the remark this answers is not in the document]
Comment (tomholub) [reply extension/chrome/elements/pgp_block.ts]: Same comment as above about it possibly fitting on one line with some changes.

File: extension/chrome/elements/pgp_block.ts [hunk 5]
```diff
@@ -436,6 +453,28 @@ Catch.try(async () => {
     }
   };
 
+  const addSecondPart = (message: string, isHtml: boolean = false) => {
+    const pgpBlock = $("#pgp_block")
+      .append('<div id="action_open_full" class="three_dots"><img src="/img/svgs/three-dots.svg" /></div>'); // xss-safe-factory
```
[the remark this answers is not in the document]
Comment (michael-volynets) [reply extension/chrome/elements/pgp_block.ts]: Okay, will do. 

Is there an article or some documentation where I can read about these comments because I can't understand exactly which xss comment I should use and where?

File: extension/chrome/elements/pgp_block.ts [hunk 6]
```diff
@@ -292,48 +292,65 @@ Catch.try(async () => {
...
         if (att.treatAs() !== 'publicKey') {
           renderableAtts.push(att);
         } else {
           publicKeys.push(att.getData().toUtfStr());
         }
       }
-      if (renderableAtts.length) {
-        renderInnerAtts(decoded.atts);
+    }
+    if (isHtml) {
```
[the remark this answers is not in the document]
Comment (michael-volynets) [reply extension/chrome/elements/pgp_block.ts]: Yes, I agree with you

File: extension/chrome/elements/pgp_block.ts [hunk 7]
```diff
@@ -292,48 +292,65 @@ Catch.try(async () => {
...
+      } else {
+        await renderContent(decryptedContent, false);
       }
-      if (publicKeys.length) {
-        BrowserMsg.send.renderPublicKeys(parentTabId, { afterFrameId: frameId, publicKeys });
+    } else {
+      const index = decryptedContent.search(/(?!^)(>.*\n?)+(\n?)+$/g);
+      if (index !== -1) {
+        await renderContent(`<div>${Xss.escapeTextAsRenderableHtml(decryptedContent.slice(0, index))}</div>`, false);
+        addSecondPart(decryptedContent.slice(index));
```
[the remark this answers is not in the document]
Comment (michael-volynets) [reply extension/chrome/elements/pgp_block.ts]: Oh, yes, I forget to rename it. It was just a test name.

File: extension/chrome/elements/pgp_block.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (michael-volynets) [reply extension/chrome/elements/pgp_block.ts]: Yes, it fits in one line

File: extension/chrome/elements/pgp_block.ts (hunk 1 above)
Comment (michael-volynets) [reply extension/chrome/elements/pgp_block.ts]: Yes, I forget about ternary operator because I don't often use it. But in this case, it looks great.
I will try to use it more often. Thanks!

File: extension/chrome/elements/pgp_block.ts (hunk 2 above)
Comment (michael-volynets) [reply extension/chrome/elements/pgp_block.ts]: Yes, I agree. I will create an issue.

It was just a habit, on previous projects that I was working on we called variables this way.

**Comment by tomholub** [comment]
Thanks, have all the comments been addressed?

File: extension/chrome/elements/pgp_block.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomholub) [reply extension/chrome/elements/pgp_block.ts]: I'll make docs about it, please make an issue for me

**Comment by michael-volynets** [comment]
Yes, I've corrected all your comments

File: extension/chrome/elements/pgp_block.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (michael-volynets) [reply extension/chrome/elements/pgp_block.ts]: Yes, I've done it.
#1752

**Comment by tomholub** [comment]
HTML doesn't seem to work on my end, or at least not the messages on the two links I provided.

Text works (see three dots).

image

image

text:
```
chrome-extension://nhlmpmiobabndkmnhocbgomijbcmiikg/chrome/elements/pgp_block.htm?frameId=frame_LXiRfsGgvp&message=-----BEGIN%20PGP%20MESSAGE-----%0AVersion%3A%20FlowCrypt%206.8.3%20Gmail%20Encryption%0AComment%3A%20Seamlessly%20send%20and%20receive%20encrypted%20email%0A%0AwcFMA0taL%2FzmLZUBAQ%2F%2FaTNLfVarEZ9e%2BMmdB8eQzsQTOJmEx%2FmRMuCgrzs9%0A91gXx%2BvwAmbJd9eU5Y8u3ca5foTl9Rf4fASEo%2BFYr9YjxcYLgJRXcMXiPc%2Fa%0ANFbIYfrKwhqnsdNk4KpghZYHwmZWBQhX%2F47264DVsyZubLgFT6YuBJyqKlLM%0ABgjdMuNnW%2Fza9JMvKN0ng%2F9QFMwV0y4Pe6XVbPSS%2BEZTZe3uLh8mWb55Mzk9%0Aqq5E%2BBFipua7xTVITc1VeqJWJgOyQfDdkczJaXDRK8cvP8W5rIgXnNzwsGAv%0AZSYBVtboDa9sMHw3YKtr0CSdRgsBAf9xFxIzulVENNRrB2e3n1oqrRSL%2FZml%0A0XaBsrnw46%2B2dzFXRAG8a1FstIGecG059GBbfreDTQAhVbawYCaKtBPdTRgg%0AKdCR6me7aqOdmty%2BoowGlEHV8KgxgcgPalUDnhezFy9lcY20W5pKHQiWPTPF%0Aghfy%2FERCzdgMvZB4%2FeowewHCFbv7%2Fpam%2FJ%2BFB%2FT1pq54CkxDPfSsnVlJG18S%0ACPIQMRpf5XZIwzc4OdmZnNRKXHah6MD46PZ8Y0JJoJ50ujdkOOq0DnYn4H6X%0AjtK%2FOeOi6tIpcQmVMgfs%2FFGRcNnyzP9k5YHlK5GwOMKMyHt1dNpmTfEedrTn%0AnX57NGwPFuqychvnQdWGpA2mGCe61onx1RzrfoNlj07BwUwDS1ov%2FOYtlQEB%0AD%2F48u6JtoqxgIoLp%2BfyfShC1k3WpviK69eXjTyBviXkaIv1YPiqH6S4iENre%0AlM%2BeTNXjIhW4p2QP1eNFNvMt4OcHPY3ANN6p1VeXsuxHEPo9OSiGikk1CuLu%0AIWNc6%2FxaXgrVhQxbjpfa7CKfoHYylV3a02W7tXwxaoAeov8Fs0PhzPKM%2BAEs%0AdIFgK6u6c5kOteMb85HloYn1I%2BtmfNb1cyao97xmGNFPJUc%2B2yQ87gI28e6U%0Ah43OYgFR%2FgAPQCapMkiDDD9TjY%2F6QR7RZoZ4ahx7eqj8pyLIPSg1yqBILGtq%0AmxigbTDiNE67DRpASFcruh0ENusje%2F15uc6bafVVdADCUEcksTVJzT%2BHChNu%0AOdG2qdreivy%2B%2BVLT7HiOl78yf%2F%2B%2FXQYApJS%2FhGHmGJFXONFHEW2N8%2FTabOV9%0AArJuT%2Bri74xMnISAoEf6Ly6Akvsr8lY0oMUC4yhqRxa4zysTnWDG2jOnr0GW%0ApEbAFu2Qz491%2FUZosuTTLptxtvKv1%2FhT%2BTT2FzyTqgLG3cnXKTWIp6SOV56i%0AxgMs1wB%2BlLi89rFjipcp2iyt1TvKLxNENT1e%2F7dd%2BLn%2FhzlS6T2BrAzUksav%0ACSt7QHvTVF6fCo7weoYfjlrxXNort3pemkz8e%2BLJAafwci5C6vSQkmAhXJtD%0A3e6ZVfNER7%2BIqlsHLUTpptgol9K%2BASRLJ%2Ftm9K1JUYSwt6Q33zvAaOyJy4SC%0AQQXX%2Bm27fE43AG2d5ZiQAwferLiH%2FeGJyG0CTOfAYGt5c4r73ca4bRzU4Bvv%0At6T556ZrheQMSEASdxTOMqeziRQ6aDB32IeX%2F2aMwIYYoDOT0jxgH1%2FzgYyP%0AeJ6qOVN%2BMO4suN0NY2%2BG2lnAxvf%2F4Rapl14ZR03pLWk7YDsiqIqrX2FlBMBt%0AWnOhe%2FHgWRDU%2BX%2FBfzEKRSObwpsmXeBNbRx11utUqQ%3D%3D%0A%3D9Bdy%0A-----END%20PGP%20MESSAGE-----&hasPassword=___cu_false___&msgId=16b584ed95837510&senderEmail=flowcrypt.compatibility%40gmail.com&isOutgoing=___cu_true___&acctEmail=flowcrypt.compatibility%40gmail.com&parentTabId=2%3A0
```

html 1:
```
chrome-extension://nhlmpmiobabndkmnhocbgomijbcmiikg/chrome/elements/pgp_block.htm?frameId=frame_GUmnyJptoI&message=&hasPassword=___cu_false___&msgId=1639b8ceb6c44a4c&senderEmail=roman.pavlovsky%40gmail.com&isOutgoing=___cu_false___&acctEmail=flowcrypt.compatibility%40gmail.com&parentTabId=2%3A0
```

html 2:
```
chrome-extension://nhlmpmiobabndkmnhocbgomijbcmiikg/chrome/elements/pgp_block.htm?frameId=frame_iIyaPcnwcQ&message=&hasPassword=___cu_false___&msgId=164563dc9e3a8549&senderEmail=franklin%40goodhorse.idv.tw&isOutgoing=___cu_false___&acctEmail=flowcrypt.compatibility%40gmail.com&parentTabId=2%3A0
```

Please also add `opacity: 0.75` to `quoted_content`

**Comment by michael-volynets** [comment]
Yes, I've fixed it.

The issue was because the last element in these emails the last element isn't `blockquote`, now it quotes  some elements that can be before the `blockquote`

**Comment by tomholub** [comment]
Maybe it's failing because the automated tests cannot now read the quoted content. I can look into it.

**Comment by michael-volynets** [comment]
Yes, it's failing because of it. 
Should I remove this content `"第四屆董事會成員、認證委員會委員"` or change it with another one?

**Comment by tomholub** [comment]
I'll make the tests click the quote button, at least it will be tested.
